### PR TITLE
ECDSA Wallets: Store uncompressed public key

### PR DESCRIPTION
**Pull request by nkuba**
We want to store a public key of the wallet so it can be used by the wallet owner (Bridge) for transactions handling. The Bridge will use both compressed and uncompressed public keys. We store the longer one (uncompressed) that can be easily converted to the compressed format.

We also renamed `publicKeyHash` to walletID as this reflects better what this property is. Using the name `publicKeyHash` was confusing in the external owner contract as there may be another key hash in use, but this is an implementation detail of what hashing function we use for identification.

Please notice that gas costs went up:

BEFORE:
|  WalletRegistry.approveDkgResult | 210844 | 225949 | 214789 │

AFTER:
|  WalletRegistry.approveDkgResult | 257700 | 257940 | 257830 |

Depends on https://github.com/keep-network/keep-core/pull/2854
Refs: https://github.com/keep-network/tbtc-v2/pull/159

### CODE REVIEW

File: solidity/ecdsa/contracts/WalletRegistry.sol [hunk 1]
```diff
@@ -482,25 +484,37 @@ contract WalletRegistry is IRandomBeaconConsumer, Ownable {
...
 
+    /// @notice Checks if a wallet with the given ID is registered.
+    /// @param walletID Wallet's ID.
+    /// @return True if wallet is registered, false otherwise.
+    function isWalletRegistered(bytes32 walletID) external view returns (bool) {
+        return wallets.isWalletRegistered(walletID);
+    }
+
+    // TODO: Add function to close the Wallet so the members are notified that
+    // they no longer need to track the wallet.
```
Comment (pdyraga) [inline solidity/ecdsa/contracts/WalletRegistry.sol:516]: In this PR or separately? If separately, please capture it in the issue and add it to [the project](https://github.com/orgs/keep-network/projects/31).

Comment (nkuba) [reply solidity/ecdsa/contracts/WalletRegistry.sol:516]: https://github.com/keep-network/keep-core/issues/2864

File: solidity/ecdsa/contracts/api/IWalletRegistry.sol [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+//   ▓▓▓▓▓▓   ▀▓▓▓▓▓▓▄     ▐▓▓▓▓▓▓     ▓▓▓▓▓   ▓▓▓▓▓▓     ▓▓▓▓▓   ▐▓▓▓▓▓▌
+// ▓▓▓▓▓▓▓▓▓▓ █▓▓▓▓▓▓▓▓▓ ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓  ▓▓▓▓▓▓▓▓▓▓
+// ▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓ ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓  ▓▓▓▓▓▓▓▓▓▓
+//
+//                           Trust math, not hardware.
+
+pragma solidity ^0.8.9;
+
+interface IWalletRegistry {
+    /// @notice Requests a new wallet creation.
```
[the remark this answers is not in the document]
Comment (nkuba) [reply solidity/ecdsa/contracts/api/IWalletRegistry.sol:18]: 0c94c4a68005d14d181b373bb152656dec2ea1ff

File: solidity/ecdsa/contracts/libraries/Wallets.sol [hunk 3]
```diff
@@ -14,22 +14,19 @@
...
     struct Wallet {
-        // TODO: Verify if we want to store the whole public key or having
-        // just the public key hash is enough.
         // Keccak256 hash of group members identifiers array. Group members do not
         // include operators selected by the sortition pool that misbehaved during DKG.
         bytes32 membersIdsHash;
-        bytes32 digestToSign;
+        // Uncompressed ECDSA public key (64-byte long) as a concatenation of X
+        // and Y coordinates.
+        bytes publicKey;
```
[the remark this answers is not in the document]
Comment (nkuba) [reply solidity/ecdsa/contracts/libraries/Wallets.sol]: > We should store the key in a compressed form 

We need both compressed and uncompressed public keys format in the Bridge to validate transactions and signatures. It's easier to convert the key from uncompressed to compressed format than the other way around. As it would require elliptic curve operations.

> use the remaining space for uint64 wallet creation date. This will be needed to manage the wallet lifecycle. Please look at wallet_creation_period and wallet_max_age parameters in [TBTCv2 RFC-1](https://github.com/keep-network/tbtc-v2/blob/main/docs/rfc/rfc-1.adoc#26-wallet-lifecycle).

We probably will want to store it in the Bridge next to the other wallet details as the Bridge will manage the wallet lifecycle as the owner.

> Have you tried to replace it with two bytes32 slots

Nice catch, it reduced cost for about 24k (fe77e91f7f5a5f81a0bc199fdc989810eb08970f)

[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/ecdsa/contracts/libraries/Wallets.sol]: Is the gas consumption in the PR description up-to-date?

[the remark this answers is not in the document]
Comment (nkuba) [reply solidity/ecdsa/contracts/libraries/Wallets.sol]: It is now.

File: solidity/ecdsa/contracts/WalletRegistry.sol [hunk 4]
```diff
@@ -390,18 +391,19 @@ contract WalletRegistry is IRandomBeaconConsumer, Ownable {
...
-        emit WalletCreated(publicKeyHash, keccak256(abi.encode(dkgResult)));
+        emit WalletCreated(walletID, keccak256(abi.encode(dkgResult)));
 
         // TODO: Disable rewards for misbehavedMembers.
         //slither-disable-next-line redundant-statements
         misbehavedMembers;
 
         walletOwner.__ecdsaWalletCreatedCallback(
-            publicKeyHash,
+            walletID,
```
Comment (pdyraga) [inline solidity/ecdsa/contracts/WalletRegistry.sol:406]: Would it make sense to return `(walletID bytes32, publicKeyX bytes32, publicKeyY bytes32)` from `wallets.addWallet` and pass `publicKeyX bytes32, publicKeyY bytes32` separately to `__ecdsaWalletCreatedCallback`?

I am looking at https://github.com/keep-network/tbtc-v2/pull/159 and I see we split `bytes` public key into x and y coordinate when compressing the key:
```
        bytes32 x = uncompressedPublicKey.slice32(0);
        bytes32 y = uncompressedPublicKey.slice32(3);
```

If `Bridge` needs separate coordinates, then maybe we can pass them to save gas on slicing?

File: solidity/ecdsa/test/WalletRegistry.Wallets.test.ts [hunk 5]
```diff
@@ -20,6 +23,183 @@ describe("WalletRegistry - Wallets", async () => {
...
+                  test.publicKey
+                )
+              ).to.be.revertedWith(test.expectedError)
+            })
+          })
+        })
+      })
+    })
+
+    context("with wallet registered", async () => {
```
Comment (pdyraga) [inline solidity/ecdsa/test/WalletRegistry.Wallets.test.ts:146]: Non-blocking but worth keeping in mind: we do not need `async` for context fn in the majority of cases.

File: solidity/ecdsa/test/WalletRegistry.Wallets.test.ts [hunk 6]
```diff
@@ -47,8 +227,54 @@ describe("WalletRegistry - Wallets", async () => {
...
+        await expect(
+          walletRegistry.getWalletPublicKey(formatBytes32String("NON EXISTING"))
+        ).to.be.revertedWith("Wallet with given ID has not been registered")
+      })
+    })
+
+    context("with wallet registered", async () => {
+      const walletPublicKey = ecdsaData.group1.publicKey
+
+      // TODO: Add tests to cover zero padding
```
Comment (pdyraga) [inline solidity/ecdsa/test/WalletRegistry.Wallets.test.ts:248]: In this PR or separately?

File: solidity/ecdsa/contracts/WalletRegistry.sol (hunk 4 above)
Comment (nkuba) [reply solidity/ecdsa/contracts/WalletRegistry.sol:406]: Yes, we can do that.

File: solidity/ecdsa/test/WalletRegistry.Wallets.test.ts (hunk 5 above)
Comment (nkuba) [reply solidity/ecdsa/test/WalletRegistry.Wallets.test.ts:146]: If it doesn't cause any problems I like that we keep it consistent and don't try to figure out where it should be placed and where it's not needed. IMHO it's more important for `await` with expects where it's not always needed but if we miss it in a place where it should be added then we will end up with a false-positive test result.

Comment (pdyraga) [reply solidity/ecdsa/test/WalletRegistry.Wallets.test.ts:146]: I might be missing something, can you give me one example in our codebase of where it is really needed?

File: solidity/ecdsa/contracts/WalletRegistry.sol (hunk 4 above)
Comment (pdyraga) [reply solidity/ecdsa/contracts/WalletRegistry.sol:406]: ... and I would leave `getWalletPublicKey` as-is because that's what we need for [`CheckBitcoinSigs` in `bitcoin-spv`](https://github.com/keep-network/bitcoin-spv/blob/solidity/v3.3.0-solc-0.8/solidity/contracts/CheckBitcoinSigs.sol#L20) that will be used by TBTC `Bridge` for fraud proofs.

File: solidity/ecdsa/test/WalletRegistry.Wallets.test.ts (hunk 5 above)
Comment (nkuba) [reply solidity/ecdsa/test/WalletRegistry.Wallets.test.ts:146]: Do you mean an example of `await` with `expect`? Here we have it, but if it wasn't provided we would not catch a test failure: https://github.com/keep-network/keep-core/blob/b6768d9bd326f4fe67bb189fafd27b5aff641c0c/solidity/ecdsa/test/WalletRegistry.Wallets.test.ts#L173-L177

I've seen numerous of times situations where we were having `expect` without an `await` where it was really necessary. This is why I prefer to add `await` to all the `expect`s, even if it's redundant.

File: solidity/ecdsa/test/WalletRegistry.Wallets.test.ts (hunk 6 above)
Comment (nkuba) [reply solidity/ecdsa/test/WalletRegistry.Wallets.test.ts]: bfb8b9b7ad89e3dd0aa61af4837f7f545c75be98

**Review by pdyraga — APPROVED** [review]
